### PR TITLE
Reworked FirefoxAccount construction

### PR DESCRIPTION
**Pull request by bendk**
Made `FirefoxAccount` input a single `FxaConfig` instance for all of it's configuration.  I want to add more constructor arguments but don't want to end up with a huge argument count.

Made the example code use the public API, not the internal module. `FxaConfig` helps ease this transition.  The motivation for this is that I want to move FxaClent and some other fields to the public `FirefoxAccount` instance, which means the example code needs to start using it. Deleted the `fxa-client/migration` example, we're not really using the migration code anymore.

Adding some handling for dead code now that we're not exposing the `internal` module as pub.

Created an `FxaServer` enum.  This matches how the android wrapper handles it and I don't think we need to input arbitrary URLs here.

### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/fxa-client/src/internal/config.rs
Comment (bendk) [inline components/fxa-client/src/internal/config.rs:166]: I feel like we should delete, but I don't know about the history of these.  What do others think?

File: components/fxa-client/src/fxa_client.udl
Comment (mhammond) [inline components/fxa-client/src/fxa_client.udl:721]: I think the fact that this exists at all is suspect tbh - I don't see why this module needs to know about the override - but that existed well before this :)

**Review by mhammond — COMMENTED** [review]
This makes sense to me, but it's a breaking change, right?

File: components/fxa-client/src/internal/config.rs
Comment (mhammond) [reply components/fxa-client/src/internal/config.rs:166]: I'm always up for killing unused code 🔥

File: components/fxa-client/src/lib.rs
Comment (bendk) [inline components/fxa-client/src/lib.rs:88]: I'm thinking that things that block like the oauth handler, API client, storage handler, etc. should live here, next to the `Mutex<internal::FirefoxAccount>`.  That would facilitate things like releasing the lock while the application performs the oauth flow.  This is why I made the CLI examples construct the public FirefoxAccount rather than the internal one.

Once those changes are made, then there's going to be a bigger distinction between `FirefoxAccount` and `internal::FirefoxAccount`.  What do others think about changing the name to reflect that?  Maybe calling it `FxaInternal` or `FxaInternalState`?

There's an alternative would be to keep `FirefoxAccount` as a thin wrapper around `internal::FirefoxAccount`, and moving a lot of functionality into the `State` struct.  I don't think I like that as much though.

**Comment by bendk** [comment]
> This makes sense to me, but it's a breaking change, right?

Yes it is.  I added the `do-not-merge` label, let's not remove it until we have approved PRs for iOS and Android.

File: components/fxa-client/src/lib.rs
Comment (tarikeshaq) [reply components/fxa-client/src/lib.rs:92]: I think it will all be clearer once we start seeing some of those changes, but here are my initial non-blocking thoughts

- By switching `internal::FirefoxAccount` into `internal::FxaInternalState`, are we going to consolidate the states into one place (putting the caches, the `State`, and the HTTP backoff state into one place and wrapping that up in the mutex?). If so I like it, in a way, we are collapsing the two `FirefoxAccount`s and making the locking more granular and +1 for that!
- How will the other members (HTTP client for example) interact with the state that requires locking the mutex? Will that be the responsibility of the high-level `FirefoxAccount`? For example after an http request if we want to update the back off for an endpoint?

File: components/fxa-client/src/fxa_client.udl
Comment (tarikeshaq) [reply components/fxa-client/src/fxa_client.udl:721]: Probably for pre-sync manager implementations, we might have an API that retrieves the token server URL for consumers

**Review by tarikeshaq — APPROVED** [review]
A couple of non-blocking comments, but this LGTM - +1 to making sure consumer PRs are up if needed (might not need one for iOS since a lot of that logic is handled here in application-services)

File: components/fxa-client/src/internal/config.rs
Comment (bendk) [reply components/fxa-client/src/internal/config.rs]: Good catch, I pushed an update for this.

File: components/fxa-client/src/lib.rs
Comment (bendk) [reply components/fxa-client/src/lib.rs:92]: My goal is to move some of the state out of `internal::FirefoxAccount`.  I'm thinking the public `FirefoxAccount` will have fields that look something like this:
  - `FxaInternalState`: holds the current `State` field, plus some new internal state, like the current state transition.
  - `ApiClient`: contains the HTTP functionality + the caches and rate limiter/backoff
  - `OAuthHandler`: The application-provided oauth handler
  - `StorageHandler`: application-provider storage handler
  - `EventHandler`, `...`: I think all other application-provided callback interfaces will probably also live here, but haven't thought all of it through.

Each of these will have its own locking, if they need it.  Let's take the `get_devices()` call as an example:
  - Call `internal_state.get_device_request()`.  That locks the state to figure out which URL, refresh token, etc. to use.  Once we know the request, we release the lock.
  - Call `api_client.execute_request(device_request)` to perform the request.
     - Part of performing the request means checking/saving to the cache which will lock a mutex internal to `ApiClient`
     - For auth-related requests it would also check the rate-limiter, which would currently require a mutex, but I think we can avoid that with a refactor
   - Finally return the result of the request, maybe after some conversion.

By having multiple top-level components, each with their own locking, we allow more concurrency.  In particular, we don't want to block on the user-interaction in OAuthhandler.  `FxaInternalState`, or whatever we end up calling it, will be able to access the internal state, but should not block on IO, network, user-interaction, etc.

Each of the components will be independent of each other so there's no chances of deadlocks.  Maybe they will need some lock-free communication channels though, for example `FxaInternalState` may need a way to signal to `ApiClient` that it should clear it's cache/reset it's rate limiter.  Maybe not though, I could see the top-level `FirefoxAccount` handling that.

File: components/fxa-client/src/fxa_client.udl
Comment (mhammond) [reply components/fxa-client/src/fxa_client.udl:721]: yeah, that might even be true today, but it still seems wrong for this to know about the override - this crate should probably just return what fxa told us, and the choice about whether to use the override made by the consumer of this

**Comment by lougeniaC64** [comment]
> > This makes sense to me, but it's a breaking change, right?
> 
> Yes it is. I added the `do-not-merge` label, let's not remove it until we have approved PRs for iOS and Android.

These changes make sense to me but I have reservations around merging this PR prior to us vetting the overall approach for this effort. I wouldn't have an issue if this were an internal refactor without breaking changes but is right on the line between refactor and the actual work to complete the FxA refactor.

I'm not opposed to these changes, but I'd feel better about introducing the breaking changes once we're more certain about the scope and strategy for this effort.

**Comment by mhammond** [comment]
> I'm not opposed to these changes, but I'd feel better about introducing the breaking changes once we're more certain about the scope and strategy for this effort.

@tarikeshaq seems to think we might be able to make it a non-breaking change by adjusting our wrappers, but the point is a really good one in general. We seem to have a tension between:

* We don't want this to be a long-running project that makes a massive PR with many breaking changes.
* We don't want to land smaller breaking changes until we have a clear picture of what we are trying to do.
The problem is I agree with *both* of them, but they seem to contradict 😅 .

File: components/fxa-client/src/fxa_client.udl
Comment (mhammond) [inline components/fxa-client/src/fxa_client.udl:732]: How do you feel about using `Url` here? As a nit, I'd say just `url` is better than `content_url`, particularly for external people who might misunderstand what `content` means in this context.

**Comment by bendk** [comment]
> These changes make sense to me but I have reservations around merging this PR prior to us vetting the overall approach for this effort. I wouldn't have an issue if this were an internal refactor without breaking changes but it's right on the line between refactor and the actual work to complete the FxA refactor.

> @tarikeshaq seems to think we might be able to make it a non-breaking change by adjusting our wrappers, but the point is a really good one in general. We seem to have a tension between:
> 
> * We don't want this to be a long-running project that makes a massive PR with many breaking changes.
> * We don't want to land smaller breaking changes until we have a clear picture of what we are trying to do.
The problem is I agree with *both* of them, but they seem to contradict 😅 .

I agree that there's a tension in general, but I think this PR is on the "okay to introduce now" side.  This actually isn't a breaking change for our consumers, I updated the wrappers to ensure that.  It's also not that much of a change from the current status quo.   The [Swift](https://github.com/mozilla/application-services/blob/main/components/fxa-client/ios/FxAClient/FxAccountConfig.swift) and [Kotlin](https://github.com/mozilla/application-services/blob/main/components/fxa-client/ios/FxAClient/FxAccountConfig.swift) already have a version of the `Config` and `Server` types that get passed to the fxa client constructor.  This is just changing the Rust code to be more inline with that.

**Comment by mhammond** [comment]
I agree - the fact it's *not* a breaking change for consumers makes this a no-brainer. I was talking more in the abstract :)

File: components/fxa-client/src/fxa_client.udl
Comment (bendk) [reply components/fxa-client/src/fxa_client.udl:732]: `url` makes more sense to me than `content_url`.  I'll make that change.

I tested out the using `Url` directly and the main issue is that parsing URLs is failing.  I think that means we need to either make the constructor failing  or panic in the face of invalid URLs.  I'm thinking we can hold off on this one for now.

**Comment by bendk** [comment]
Updated the wrapper code to remove the breaking change.  I think I applied all the suggestions except switching to `Url`, which seems better as a follow-up.  The CI is green.  Should we merge this one?

File: components/fxa-client/src/lib.rs
Comment (mhammond) [inline components/fxa-client/src/lib.rs:43]: I think we fixed that so it's no longer true. I put a PR up for your branch to kill this!

**Review by mhammond — APPROVED** [review]
This looks great, thanks! I think you might as well fold my PR into this too assuming it's green. Or as a followup, I don't really mind, but it's logically a part of this.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/Config.kt
Comment (mhammond) [inline components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/Config.kt:17]: One interesting issue with this refactor is that it seems impossible for consuming apps to discover what the actual URL being used it? ie, they might construct with `FxaServer.Stage`, but later when they want to open, say, the "manage account" page, they seem unable to discover what the base URL is. Before this change they are able to obtain it via the enum.

Not sure what the answer is here - ie, whether we should expose the base URL, or expose a way to build a URL - eg, how desktop does it is [here](https://searchfox.org/mozilla-central/source/services/fxaccounts/FxAccountsConfig.sys.mjs#79) - the code is responsible for knowing whether the email address should be included in the URL etc.

File: components/fxa-client/src/lib.rs
Comment (bendk) [reply components/fxa-client/src/lib.rs]: Sounds great to me.  I just merged the branch in.

**Pull request merged**
